napalm_ios/ios.py

```diff
@@ -1439,6 +1439,123 @@ def ping(self, destination, source='', ttl=255, timeout=2, size=100, count=5):
+        max_loops = (5 * ttl_value * timeout_value) + 150
+        output = self.device.send_command(command, max_loops=max_loops)
+
+        # Prepare return dict
+        traceroute_dict = dict()
+        if re.search('Unrecognized host or address', output):
+            traceroute_dict['error'] = 'unknown host %s' % destination
+            return traceroute_dict
+        else:
+            traceroute_dict['success'] = dict()
+
+        results = dict()
+        # Find all hops
+        hops = re.findall('\\n\s+[0-9]{1,3}\s', output)
+        for hop in hops:
+            # Search for hop in the output
+            hop_match = re.search(hop, output)
+            # Find the start index for hop
+            start_index = hop_match.start()
+            # If this is last hop
+            if hops.index(hop) + 1 == len(hops):
+                # Set the stop index for hop to len of output
+                stop_index = len(output)
+            # else, find the start index for next hop
+            else:
+                next_hop_match = re.search(hops[hops.index(hop) + 1], output)
+                stop_index = next_hop_match.start()
+                # Now you have the start and stop index for each hop
+                # and you can parse the probes
+            # Set the hop_variable, and remove spaces between msec for easier matching
+            hop_string = output[start_index:stop_index].replace(' msec', 'msec')
+            hop_list = hop_string.split()
+            current_hop = int(hop_list.pop(0))
+            # Prepare dictionary for each hop (assuming there are 3 probes in each hop)
+            results[current_hop] = dict()
+            results[current_hop]['probes'] = dict()
+            results[current_hop]['probes'][1] = {'rtt': float(),
+                                                 'ip_address': unicode(),
+                                                 'host_name': unicode()}
+            results[current_hop]['probes'][2] = {'rtt': float(),
+                                                 'ip_address': unicode(),
+                                                 'host_name': unicode()}
+            results[current_hop]['probes'][3] = {'rtt': float(),
+                                                 'ip_address': unicode(),
+                                                 'host_name': unicode()}
+            current_probe = 1
+            ip_address = ''
+            host_name = ''
+            while hop_list:
+                current_element = hop_list.pop(0)
+                # If current_element is * move index in dictionary to next probe
+                if current_element == '*':
+                    current_probe += 1
+                # If current_element contains msec record the entry for probe
+                elif 'msec' in current_element:
+                    rtt = float(current_element.replace('msec', ''))
+                    results[current_hop]['probes'][current_probe]['ip_address'] = ip_address
```

Comment:
> I should use this py23_compat.text_type(ip_address) when adding the ip_address to the dictionary?

Yes, and also for ```host_name```.

> Or is it enough to change unicode to ``` '' ```?

You should do that too, otherwise Python3 tests will fail.